flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/Bucket.java

```diff
@@ -0,0 +1,297 @@
+
+		this(fsWriter, subtaskIndex, bucketPath, initialPartCounter, maxPartSize, rolloverTime, inactivityTime, writer);
+
+		// the constructor must have already initialized the filesystem writer
+		Preconditions.checkState(fsWriter != null);
+
+		// we try to resume the previous in-progress file, if the filesystem
+		// supports such operation. If not, we just commit the file and start fresh.
+
+		final ResumableWriter.ResumeRecoverable resumable = bucketstate.getCurrentInProgress();
+		if (resumable != null) {
+			this.currentOpenPartStream = fsWriter.recover(resumable);
+			this.creationTime = bucketstate.getCreationTime();
+		}
+
+		// we commit pending files for previous checkpoints to the last successful one
+		// (from which we are recovering from)
+		for (List<ResumableWriter.CommitRecoverable> commitables: bucketstate.getPendingPerCheckpoint().values()) {
+			for (ResumableWriter.CommitRecoverable commitable: commitables) {
+				fsWriter.recoverForCommit(commitable).commit();
+			}
+		}
+
+		this.pending = new ArrayList<>();
+		this.pendingPerCheckpoint = new HashMap<>();
+	}
+
+	public Bucket(
+			ResumableWriter fsWriter,
+			int subtaskIndex,
+			Path bucketPath,
+			long initialPartCounter,
+			long maxPartSize,
+			long rolloverTime,
+			long inactivityTime,
+			Writer<IN> writer) {
+
+		this.fsWriter = Preconditions.checkNotNull(fsWriter);
+		this.subtaskIndex = subtaskIndex;
+		this.bucketPath = Preconditions.checkNotNull(bucketPath);
+		this.partCounter = initialPartCounter;
+		this.maxPathSize = maxPartSize;
+		this.rolloverTime = rolloverTime;
+		this.inactivityTime = inactivityTime;
+		this.outputFormatWriter = Preconditions.checkNotNull(writer);
+		this.creationTime = Long.MAX_VALUE;
+	}
+
+	public Path getBucketPath() {
+		return bucketPath;
+	}
+
+	public long getPartCounter() {
+		return partCounter;
+	}
+
+	public boolean isActive() {
+		return isOpen() || !pending.isEmpty() || !pendingPerCheckpoint.isEmpty();
+	}
+
+	public boolean isOpen() {
+		return currentOpenPartStream != null;
+	}
+
+	public void write(IN element, Long timestamp, long currentTime) throws IOException {
+		if (shouldRoll(currentTime)) {
+			startNewChunk(currentTime);
+		}
+		Preconditions.checkState(isOpen(), "BucketingSink is not open.");
+
+		outputFormatWriter.write(element, timestamp, currentOpenPartStream);
+		lastWrittenTime = currentTime;
+	}
+
+	private void startNewChunk(final long currentTime) throws IOException {
+		closeCurrentChunk();
+
+		this.currentOpenPartStream = fsWriter.open(getNewPartPath());
+		this.creationTime = currentTime;
+		this.partCounter++;
+	}
+
+	private boolean shouldRoll(long currentTime) throws IOException {
+		if (currentOpenPartStream == null) {
+			return true;
+		}
+
+		long writePosition = currentOpenPartStream.getPos();
+		if (writePosition > maxPathSize) {
+			return true;
+		}
+
+		return (currentTime - creationTime > rolloverTime);
+	}
+
+	void merge(final Bucket bucket) throws IOException {
```

Comment:
This should probably assert that the input bucket actually doesn't have any pending `CommitRecoverables`, otherwise we will lose them here.

Reply:
You are right that we have to have a `Precondition` here but the `pending` ones should be committed by the constructor of the `Bucket` with the `BucketState`.